File: suanfa/quick-sort/quick_sort.c
//
//  quick_sort.c
//  mynapi
//
//  Created by tangxiaoji on 2020/3/19.
//  Copyright © 2020 tangxiaoji. All rights reserved.
//

#include "quick_sort.h"
// #include <node_api.h>

typedef struct _Range {
    int start, end;
} Range;

Range new_Range(int s, int e) {
    Range r;
    r.start = s;
    r.end = e;
    return r;
}

void swap(int *x, int *y) {
    int t = *x;
    *x = *y;
    *y = t;
}

void quick_sort(int arr[], const int len) {
    if (len <= 0)
        return;
    Range r[len];
    int p = 0;
    r[p++] = new_Range(0, len - 1);
    while (p) {
        Range range = r[--p];
        if (range.start >= range.end)
            continue;
        int mid = arr[(range.start + range.end) / 2];
        int left = range.start, right = range.end;
        do
        {
            while (arr[left] < mid) ++left;
            while (arr[right] > mid) --right;
 
            if (left <= right)
            {
                swap(&arr[left],&arr[right]);
                left++;
                right--;
            }
        } while (left <= right);
 
        if (range.start < right) r[p++] = new_Range(range.start, right);
        if (range.end > left) r[p++] = new_Range(left, range.end);
    }
}

File: suanfa/quick_sort.js
'use strict'

// 快排
function quickSort(arr, i, j) {
    if (i < j) {
        let left = i, right = j, p = arr[left]
        while(i < j) {
            while(arr[j] >= p && i < j) {
                j--
            }
            if (i < j) {
                arr[i++] = arr[j]
            }
            while(arr[i] <= p && i < j) {
                i++
            }
            if (i < j) {
                arr[j--] = arr[i]
            }
        }
        arr[i] = p
        quickSort(arr, left, i - 1)
        quickSort(arr, i + 1, right)
        return arr
    }
}

const assert = require('assert')
describe('Test', () => {
    it('1', () => {
        const arr = [2, 10, 4, 1, 0, 9, 5 ,2];
        assert.deepEqual(quickSort(arr, 0 , arr.length-1), [ 0, 1, 2, 2, 4, 5, 9, 10 ])
    })
})
